### DISCUSSION TITLE
Help to configure the tree shaking

### DISCUSSION THREAD

**Discussion by JBustin** [nuxt/nuxt#10455] [Questions (Nuxt 2)] [answered]
Hello community,

I'm trying to find a way to effectively use the tree shaking with nuxtjs.

The sideEffect seems to work with no additional config (thanks webpack4), but usedExports not seem to work.
Except If I didn't understand the purpose of this features.

To resume, I'm experimenting two problems:

**Dead code using ES6 import/export si removed, for this case:**

`~/common/utils.js`
```js
export function foo() {}
export function bar() {}
```

`~/moduleA/index.js`
```js
import { foo } from '~/common/utils'
// use foo
foo()
```

=> bar is not in the bundle including A 👍 

**But it is not efficient for this case:**

`~/moduleA/index.js` 
```js
import { foo } from '~/common/utils'
foo()
```

`~/moduleB/index.js`
```js
import { bar } from '~/common/utils'
bar()
```

=> bar and foo are including in both bundles A and B 👎 

Note that each module is dynamically imported and has a dedicated bundle.
So, if my common is sharing between modules, all exports (used or not) are in the two bundles, which is not very optimized...

**Second problem:**

My app bundled and compressed size is **1.33M**.
I have tried to apply some specific tree shaking rules on js:

```js
config.module.rules.push(
    {
      test: /\.js$/,
      exclude: [
        path.resolve(__dirname, '../.nuxt'),
        path.resolve(__dirname, '../src/config'),
        path.resolve(__dirname, '../src/app'),
        path.resolve(__dirname, '../src/router.js'),
        path.resolve(__dirname, '../nuxt.config.js'),
      ],
      sideEffects: false,
    }
  )
}
```

I have no real explanation, but my app size is now **1.26M**.

My app is composed to a huge base of single file components.
If I apply some additional rules to apply the tree shaking on vue file like this:

```js
config.module.rules.push(
    {
      test: /\.js$/,
      exclude: [
        path.resolve(__dirname, '../.nuxt'),
        path.resolve(__dirname, '../src/config'),
        path.resolve(__dirname, '../src/app'),
        path.resolve(__dirname, '../src/router.js'),
        path.resolve(__dirname, '../nuxt.config.js'),
      ],
      sideEffects: false,
    },
    {
      test: /\.vue$/,
      sideEffects: false,
    }
  )
}
```

Webpack removes the `style lang=scss` section of the component, I'm loosing the internal component css.
I have tried a lot of thing, but without any success.
Only adding this has resolved the issue, but the bundles are slightly larger (globally + 10ko ?) to **1.27M**:

```js
      config.module.rules.push(
        {
          test: /\.js$/,
          exclude: [
            path.resolve(__dirname, '../.nuxt'),
            path.resolve(__dirname, '../src/config'),
            path.resolve(__dirname, '../src/app'),
            path.resolve(__dirname, '../src/router.js'),
            path.resolve(__dirname, '../nuxt.config.js'),
          ],
          sideEffects: false,
        },
        {
          test: /\.vue$/,
          sideEffects: false,
        },
        {
          test: /\.vue$/,
          resourceQuery: /lang=(s)css/,
          sideEffects: true,
        },
      )
```

So, it's very confusing for me.
Do someone has succeed to configure an optimized tree shaking with nuxtjs ?

Regards

**Comment by gplusdotgr** [comment]
Hi @JBustin 

> But it is not efficient for this case:
> 
> ~/moduleA/index.js
> 
> import { foo } from '~/common/utils'
> foo()
> ~/moduleB/index.js
> 
> import { bar } from '~/common/utils'
> bar()
> => bar and foo are including in both bundles A and B 👎

I suspect there's no split chunk for this piece of code, thats why it ends up in both files. Nuxt has some built-in rules to split chunks for big stuff (vendors, commons etc) but if you want to explicitely split specific thinks in separate chunks you'd need to add that config yourself. This is the nuxt default config for this:
```js
    splitChunks: {
      chunks: 'all',
      automaticNameDelimiter: '/',
      cacheGroups: {
        commons: {
          test: /node_modules[\\/](vue|vue-loader|vue-router|vuex|vue-meta|core-js|@babel\/runtime|axios|webpack|setimmediate|timers-browserify|process|regenerator-runtime|cookie|js-cookie|is-buffer|dotprop|url-polyfill|@nuxt[\\/]ufo|ufo|nuxt\.js)[\\/]/,
          chunks: 'all',
          name: true,
          priority: 10
        }
      }
```

doc: https://nuxtjs.org/docs/configuration-glossary/configuration-build#splitchunks 

Could you paste your original code or try to make a minimal reproduction in https://codesandbox.io/ ?
In order to have a named chunk and be able to properly inspect it, if you use dynamic imports:  
` IconCompany: () => import('@/components/icons/company-placeholder.vue')`
be sure to use `magic comments` ( https://webpack.js.org/api/module-methods/#magic-comments ) , specifically webpackChunkName . 

In our config we have added the following, to make sure our chunks have the component name in the output file: 

```js
  build: {
    filenames: {
      chunk: ({ isDev, isModern }) =>
        isDev ? `[name]${isModern ? '.modern' : ''}.js` : `[name]-[contenthash:7]${isModern ? '.modern' : ''}.js`,
    },
  }
```

**Comment by JBustin** [comment]
Hello @gkatsanos, 

Thank you for your response.
I have created a git repository with some basic code, exactly what I do and what I get from my project (see the readme.md to test)

https://github.com/JBustin/nuxt-tree-shaking

I have the same result as I have described previously.

To resume:
- A plugin to dynamically load some modules https://github.com/JBustin/nuxt-tree-shaking/blob/main/plugins/module-loader.js
- Two modules, for each the "entrypoint" is a controller file https://github.com/JBustin/nuxt-tree-shaking/blob/main/modules/a/controller.js
- A lib shared between the two modules https://github.com/JBustin/nuxt-tree-shaking/blob/main/common/shared-lib.js

What I expect:
- From the two dynamic imports inside the plugin, webpack builds two chunks 👍 
- Each chunk includes the controller and the other static deps 👍 
- Each chunk includes only the needed exports from the shared lib 👎 

What I assume:
- I have tested with the magic comments from webpack, same result
- Webpack nuxt default config is sufficient for this basic tree shaking
- If it's not, I can add some additional config, no problem

Great thanks for the discussion, and eventually for your ideas to go further.

**Reply by danielroe** [reply, marked as answer]
I think it might be helpful to check out [this page](https://v4.webpack.js.org/plugins/split-chunks-plugin/), which may explain why it’s not splitting the chunks at the moment. As @gkatsanos says, this is likely about the configuration of the split chunks plugin.

A side note - rather than pushing additional rules, you can modify the existing ones - see [loaders configuration](https://nuxtjs.org/docs/configuration-glossary/configuration-build#loaders) or (a bit more complex) [an example of finding and modifying a loader](https://nuxtjs.org/docs/directory-structure/modules/#register-custom-webpack-loaders).

**Reply by JBustin** [reply]
Thanks for the share, simple hypothesis would be that chunk is too small https://v4.webpack.js.org/plugins/split-chunks-plugin/#defaults to be optimized.

I have effectively tried by updating the minSize, and I get a different result with a very low value, the shared lib has its own chunk.
I guess for the next weeks / months, that, with increasing code for this omptimized part of the application, webpack will build some optimized chunks, and I will keep an eye on it. 🫡 

By the way, we will try to move to the nuxt-edge, probably some optimizations will become useless.

Regards
